### ISSUE TITLE
Underscores in numeric literals not supported in lib2to3.

### ISSUE THREAD

**Issue by bcc3af7a-4f73-4e81-80e1-a2a4b7370875** [python/cpython#74055]
BPO | [29869](https://bugs.python.org/issue29869)
--- | :---
Nosy | @birkenfeld, @serhiy-storchaka, @Mariatta, @nevsan
PRs | <li>python/cpython#752</li><li>python/cpython#1109</li><li>python/cpython#1119</li><li>python/cpython#1122</li><li>python/cpython#1125</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-04-14.01:31:24.462>
created_at = <Date 2017-03-22.00:12:25.319>
labels = ['3.7', 'type-bug', 'expert-2to3']
title = 'Underscores in numeric literals not supported in lib2to3.'
updated_at = <Date 2017-04-14.01:31:24.461>
user = 'https://github.com/nevsan'
```

bugs.python.org fields:
```python
activity = <Date 2017-04-14.01:31:24.461>
actor = 'Mariatta'
assignee = 'none'
closed = True
closed_date = <Date 2017-04-14.01:31:24.462>
closer = 'Mariatta'
components = ['2to3 (2.x to 3.x conversion tool)']
creation = <Date 2017-03-22.00:12:25.319>
creator = 'nevsan'
dependencies = []
files = []
hgrepos = []
issue_num = 29869
keywords = []
message_count = 14.0
messages = ['289951', '289982', '289988', '290001', '290007', '290008', '290023', '291598', '291600', '291618', '291623', '291634', '291635', '291636']
nosy_count = 4.0
nosy_names = ['georg.brandl', 'serhiy.storchaka', 'Mariatta', 'nevsan']
pr_nums = ['752', '1109', '1119', '1122', '1125']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue29869'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by bcc3af7a-4f73-4e81-80e1-a2a4b7370875** [comment]
The following should work in Python 3.6

```
from lib2to3.pgen2 import driver
from lib2to3 import pytree
from lib2to3 import pygram

_GRAMMAR_FOR_PY3 = pygram.python_grammar_no_print_statement.copy()
parser_driver = driver.Driver(_GRAMMAR_FOR_PY3, convert=pytree.convert)
tree = parser_driver.parse_string('100_1\n', debug=False)
```

**Comment by serhiy-storchaka** [comment]
Python uses more strong rules for underscores in numerical literals. Underscores are acceptable between digits and between the base prefix and digit. For example the regular expression for hexadecimals is r'0[xX]_?[\\da-fA-F]+(?:_[\\da-fA-F]+)*[lL]?'.

Underscores also are acceptable in the exponent of float literals:

```py
Exponent = r'[eE][-+]?\d+(?:_\d+)*'
```

**Comment by bcc3af7a-4f73-4e81-80e1-a2a4b7370875** [comment]
The existing regular expressions weren't actually strict enough as is, so I made them even more correct. In particular, we must have at least one digit following `0[xXbBoO]` and must be before any underscores.

I have a small set of test cases to examine correctness of these regular expressions: https://gist.github.com/nevsan/7fc78dc61d309842406d67d6839b9861

**Comment by birkenfeld** [comment]
> In particular, we must have at least one digit following `0[xXbBoO]` and must be before any underscores.

This is not true (but your test file does the right thing).

**Comment by bcc3af7a-4f73-4e81-80e1-a2a4b7370875** [comment]
Thanks, it seems I misspoke. Glad I tested it!

**Comment by serhiy-storchaka** [comment]
I suggest to use my regular expression for haxadedecimals. Your regular expression starves from catastrophic backtracking. Compare two examples:

re.match(r'0[xX]_?[\\da-fA-F]+(?:_[\\da-fA-F]+)*[lL]?'+r'\\b', '0x'+'0'\*100+'z')
re.match(r'0[xX](?:_?[\\da-fA-F]+)+[lL]?'+r'\\b', '0x'+'0'\*100+'z')

**Comment by bcc3af7a-4f73-4e81-80e1-a2a4b7370875** [comment]
Good point. Updated.

**Comment by Mariatta** [comment]
New changeset 97a40b7a5b2979fb17e1751c139fd4ba1ebd5276 by Mariatta (Nevada Sanchez) in branch '3.6':
bpo-29869: Allow underscores in numeric literals in lib2to3. (GH-752)
https://github.com/python/cpython/commit/97a40b7a5b2979fb17e1751c139fd4ba1ebd5276

**Comment by Mariatta** [comment]
New changeset 84c2d75489a84174d8993aea292828662e35a50f by Mariatta in branch '3.6':
Revert "bpo-29869: Allow underscores in numeric literals in lib2to3. (GH-752)" (GH-1109)
https://github.com/python/cpython/commit/84c2d75489a84174d8993aea292828662e35a50f

**Comment by Mariatta** [comment]
Nevada Sanchez, please create your PR against the master branch.
Once that is merged, we will backport it to the 3.6 branch.

Thanks :)

**Comment by Mariatta** [comment]
New changeset a6e395dffadf8c5124903c01ad69fefa36b1a935 by Mariatta (Nevada Sanchez) in branch 'master':
bpo-29869: Allow underscores in numeric literals in lib2to3. (GH-1119)
https://github.com/python/cpython/commit/a6e395dffadf8c5124903c01ad69fefa36b1a935

**Comment by Mariatta** [comment]
New changeset 2cdf087d1fd48f7d0f95b5a0b31b9a624fa84751 by Mariatta in branch '3.6':
[3.6] bpo-29869: Allow underscores in numeric literals in lib2to3. (GH-1119) (GH-1122)
https://github.com/python/cpython/commit/2cdf087d1fd48f7d0f95b5a0b31b9a624fa84751

**Comment by Mariatta** [comment]
New changeset 947629916a5ecb1f6f6792e9b9234e084c5bf274 by Mariatta in branch 'master':
bpo-29869: Add Nevada Sanchez to [Misc/ACKS](https://github.com/python/cpython/blob/main/Misc/ACKS) (GH-1125)
https://github.com/python/cpython/commit/947629916a5ecb1f6f6792e9b9234e084c5bf274

**Comment by Mariatta** [comment]
PR has been merged and backported to 3.6.
I also added Nevada Sanchez to [Misc/ACKS](https://github.com/python/cpython/blob/main/Misc/ACKS).
Thanks all :)
